### ISSUE TITLE
IO::Scalar::print does not take under account the position inside the data and do a mere appending

### ISSUE THREAD

**Issue by jackdeguest** [genio/IO-Stringy#4]
Hello,

As of version 2.113, `print` inside `IO::Scalar` merely append data provided and does not take under account the position inside the scalar:

    *$self->{Pos} = length(${*$self->{SR}} .= join('', @_) . (defined($\) ? $\ : ""));

To demonstrate this, consider this script:

    #!/usr/bin/perl
    BEGIN
    {
        use strict;
        use warnings;
        use open ':std' => ':utf8';
        use IO::Scalar;
    };

    {
        use utf8;
        my $text = <<EOT;
    Mignonne, allons voir si la rose
    Qui ce matin avoit desclose
    Sa robe de pourpre au Soleil,
    A point perdu cette vesprée
    Les plis de sa robe pourprée,
    Et son teint au vostre pareil.

    Author: Pierre de Ronsard
    EOT
        open( my $fh, '+>/tmp/test.txt' );
        $fh->binmode( ':utf8' );
        $fh->print( $text );
        $fh->seek( length( $text ) - 1, 0 );
        $fh->print( ', Les Odes' );
        $fh->close;
    
        # Now trying with IO::Scalar
        my $data;
        my $s = IO::Scalar->new( \$data );
        $s->print( $text );
        $s->seek( length( $text ) - 1, 0 );
        $s->print( ', Les Odes' );
        print( "Scalar is now:\n$s" );
        $s->close;
    }

    __END__

The file `/tmp/test.txt` will contain, as expected:

    Mignonne, allons voir si la rose
    Qui ce matin avoit desclose
    Sa robe de pourpre au Soleil,
    A point perdu cette vesprée
    Les plis de sa robe pourprée,
    Et son teint au vostre pareil.

    Author: Pierre de Ronsa, Les Odes

Whereas the script that prints out the resulting value of IO::Scalar would yield:

    Scalar is now:
    Mignonne, allons voir si la rose
    Qui ce matin avoit desclose
    Sa robe de pourpre au Soleil,
    A point perdu cette vesprée
    Les plis de sa robe pourprée,
    Et son teint au vostre pareil.

    Author: Pierre de Ronsard
    , Les Odes

The expectation is that the additional text printed to the scalar should have been printed before the last line feed as positioned earlier with the `seek`.

I am running perl v5.32 under Ubuntu 20.04

[Scalar.pm.txt](https://github.com/genio/IO-Stringy/files/6526127/Scalar.pm.txt)

Attached is a diff file with my proposed correction. I had to set the file suffix to .txt because Git would not let me use .diff

**Comment by genio** [comment]
The diff `.txt` file above has the contents:
```diff
429c429,431
<     *$self->{Pos} = length(${*$self->{SR}} .= join('', @_) . (defined($\) ? $\ : ""));
---
>     my $len  = length( ${*$self->{SR}} );
>     substr( ${*$self->{SR}}, *$self->{Pos}, 0, join( '', @_ ) . (defined( $\ ) ? $\ : "" ) );
>     *$self->{Pos} += ( length( ${*$self->{SR}} ) - $len );
```

**Comment by genio** [comment]
Hi @jackdeguest
Was this working in any prior version?

No code was changed in IO::Scalar between 2.112 and 2.113. No code changes were made for 2.112.  I've gone back to 2.110 and haven't found any changes to that method.

Granted, this isn't a sticking point, I'm just confused because you stated that it's been broken since 2.113 so I'm trying to understand.

**Comment by jackdeguest** [comment]
Correction, I did not state it was broken __since__ version 2.113, but rather that "as of version 2.113" it is broken.

**Comment by haarg** [comment]
Why are you using IO::Scalar at all? It has been irrelevant since perl 5.8.

**Comment by jackdeguest** [comment]
Because you cannot do this:

```perl
my $str = 'hello';
my $scalar_object = MyScalar->new( \$str );
my $io = IO::File->new( $scalar_object, '>:' );
# this returns no such file or directory. It only works with non-object scalar reference.

package MyScalar;
sub new { return( bless( $_[1] => ( ref( $_[0] ) || $_[0] ) ) ); }
```

**Comment by Leont** [comment]
> Because you cannot do this

What are you trying to achieve here? I'm not sure I see your point.

**Comment by jackdeguest** [comment]
> What are you trying to achieve here? I'm not sure I see your point.

You cannot pass a blessed scalar reference to perl's `open` it would not work.

**Comment by Leont** [comment]
> You cannot pass a blessed scalar reference to perl's open it would not work.

But why would you want to do that? I can't think of any use of such a thing.

**Comment by jackdeguest** [comment]
> But why would you want to do that? I can't think of any use of such a thing.

Obviously, if one has a scalar reference based class and wants to get an I/O interface to mimic read/write to it like an ordinary glob.

**Comment by Leont** [comment]
That doesn't really answer my question, but as far as I can tell your problem disappears if you replace that `'>:'` with `'>:scalar'`

**Comment by jackdeguest** [comment]
Thank you. Yes, using `>:scalar` solved it.

**Comment by jackdeguest** [comment]
Still, using perl `open` via IO::Handle on a scalar reference has some drawbacks, such as with `truncate`, which does not work. It would produce `Bad file descriptor`

**Comment by Leont** [comment]
> Still, using perl open via IO::Handle on a scalar reference has some drawbacks, such as with truncate, which does not work. It would produce Bad file descriptor

Yeah that is a bug, it might even have already been reported.
